### PR TITLE
Add XMLOutputParser

### PR DESCRIPTION
**Pull request by mateusz-wosinski-ds**
**Description**
Adds new output parser, this time enabling the output of LLM to be of an XML format. Seems to be particularly useful together with Claude model. 
Addresses [issue 9820](https://github.com/langchain-ai/langchain/issues/9820).

**Twitter handle**
@deepsense_ai @matt_wosinski

### CODE REVIEW

File: libs/langchain/tests/unit_tests/output_parsers/test_xml_parser.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@pytest.mark.parametrize("result", [DEF_RESULT_ENCODING, DEF_RESULT_EXPECTED])
+def test_xml_output_parser(result: str) -> None:
+    """Test XMLOutputParser."""
+
+    xml_parser = XMLOutputParser()
+
+    result = xml_parser.parse(result)
+    print("parse_result:", result)
```
Comment (hwchase17) [inline libs/langchain/tests/unit_tests/output_parsers/test_xml_parser.py:29]: stray print statement

**Review by hwchase17 — COMMENTED** [review]
interesting so this still parses it to a single string? wouldnt it make more sense to parse to a list of dictionaries or something? eg the json loader doesnt return a json string, it returns the result of calling `json.load`

**Comment by grabbou** [comment]
Thanks @mateusz-wosinski-ds for working on this PR! Awesome stuff! I guess I better get started with the JS one then!

**Comment by grabbou** [comment]
> interesting so this still parses it to a single string? wouldnt it make more sense to parse to a list of dictionaries or something? eg the json loader doesnt return a json string, it returns the result of calling `json.load`

@hwchase17, I initially proposed in #9820 parsing to a single string according to the examples featured in the Claude documentation. The response is typically a mix of XML tags and text. Typically, you would want each item to be represented by a separate XML tag rather than entire response with multiple items included in a big large `<response>` tag. Here's an example:
```
Here is the answer to your question: <answer>...</answer>
Also, here is the second thing that you requested: <analysis>...</analysis> 
```

In that case, I would expect the return of the LLM call to be an object with `answer` and `analysis` keys (assuming these are the keys that I specified while configuring the XML Parser).

We should, however, handle one additional case when multiple instances of the same tag are present in the response, e.g.
```
Here is my response to your first question: <response></response>
Here is my response to your second question: <response></response>
```

I think this would be a very natural Claude way to parse and treat multiple items (lists).

**Comment by hwchase17** [comment]
hmmm - @eyurtsev any thoughts on proper way to do parsing? i know you did this a bunch for kor

**Comment by mateusz-wosinski-ds** [comment]
I believe the way of parsing @grabbou suggests makes sense for Claude, but it does not go in line with the general idea of output parsers which should return **structured** response for any given model. The way it is implemented now, we get structured response every time and can modify it later on according to our needs. Also, as it is shown in the notebook, we can ommit parsing, and leave the output as an XML string.

@grabbou if these changes do not address your issue, then I think it should be developed in a separate PR :raised_hands:

**Comment by piotr-grodek-dsai** [comment]
In my opinion it would be good to have both and at first glance and with current implementation and design it seems lines 28-29 contain the logic which changes str into dict (please correct me if I am wrong and string output differs for input to parse as XML). I dislike idea of doing unnecessary work and conversion for me to dict looks like an adapter.

I see a few solutions:

**a) Keep XMLOutputParser and have a param to determine return type**
However above are non-standard behavior as rather my understanding the design is that parser has clearly defined output type.

**b) have an adapter function/method parse_to_dict() /to_dict or similar**
This is opt-in but results in discoverability issue, as it might be "surprising" for developers. Potentially string_xml_to_dict is reusable in other scenarios (but not sure if any..)

**c) distinct parsers - it is not much work really.**

`XMLOutputParserIntoString` it could be a base class or just implementation detail for `XMLOutputParserIntoDict` which can reuse the first parsing behavior.
It is easy to add output as list etc, typing is verbose and clear.
Maybe the name should be also more like DictFromXMLOutputParser idk. The downside of course is that runtime output format change is not easy and more verbose naming, plus tiny bit of additional work.

Personally I think option c would be the best as it gives more power to end users - what do you think?

**Comment by mateusz-wosinski-ds** [comment]
@hwchase17 any news on that?

**Review by hwchase17 — APPROVED** [review]

**Pull request merged**